sw/SoC/examples/blinky/Makefile

```diff
@@ -1,81 +1,84 @@
+#	Currently, only rv32 is supported.
 
-PROJECT = filename
-MCU = DEMOSYSTEM
+################
+# Program Name #
+################
 
-###############################################################################
+PROGRAM_NAME = $(shell basename "$$PWD")
 
-###############################################################################
-# Directories
+#####################
+# Paths and Folders #
+#####################
 
-SOC_SW_ROOT_DIR = ../..
-SRCS = 		      $(wildcard src/*.c) $(wildcard lib/*.c)
-OBJ = 		      obj/
-OBJS = 		      $(addprefix $(OBJ), $(notdir $(SRCS:.c=.o)))
-LD_SCRIPT =       $(SOC_SW_ROOT_DIR)/linker/$(MCU)_FLASH.ld
-INC = 		      -Iinc
-STARTUP = 	      startup
+SOC_SW_ROOT_DIR = $(SW_ROOT)/SoC
 
-LIBLIST = 
+SRC_DIR		= src
+OBJ_DIR		= obj
+INC_DIR 	= inc 
+LD_SCRIPT 	= ld/user.ld
+STARTUP_DIR = $(SOC_SW_ROOT_DIR)/startup
+LIB_DIR		= $(SOC_SW_ROOT_DIR)/lib
 
-###############################################################################
-# Toolchain
+#############
+# Libraries #
+#############
 
-# If you have 64-bit toolchain
-# RV_PREFIX = riscv64-unknown-elf-
-# 32BITCCFLAGS = -march=rv32im -mabi=ilp32
-# 32BITODFLAGS = -mriscv:rv32
-# CC = $(RV_PREFIX)gcc $(32BITCCFLAGS)
-# OBJDUMP = $(RV_PREFIX)objdump $(32BITODFLAGS)
-#
+LIB_OBJ_LIST 	= 
+LIB_INC_LIST 	= 
 
-RV_PREFIX = 	riscv32-unknown-elf-
-CC = 			$(RV_PREFIX)gcc
-LD = 			$(RV_PREFIX)ld
-OBJDUMP = 		$(RV_PREFIX)objdump
-OBJCOPY = 		$(RV_PREFIX)objcopy
+#############
+# Toolchain #
+#############
 
-CFLAGS = 		-march=rv32imad_zicsr_zifencei -O0 -c 
-LDFLAGS = 		$(LIBLIST) -nostdlib -T$(LD_SCRIPT)
+RV_PREFIX 	= riscv32-unknown-elf-
+CC 			= $(RV_PREFIX)gcc
+LD 			= $(RV_PREFIX)ld
+OBJDUMP 	= $(RV_PREFIX)objdump
+OBJCOPY 	= $(RV_PREFIX)objcopy
 
+CFLAGS 		= -march=rv32imad_zicsr_zifencei -O0 -c 
+LDFLAGS 	= $(LIB_OBJ_LIST) -nostdlib -T$(LD_SCRIPT)
 
 
-###############################################################################
+########
+# Misc #
+########
 
-###############################################################################
-# Misc
+SRCS 	= $(wildcard src/*.c) $(wildcard lib/*.c)
+OBJS 	= $(addprefix $(OBJ_DIR)/, $(notdir $(SRCS:.c=.o)))
 
 RM      = rm -rf 					# Remove recursively command
 MKDIR   = @mkdir -p $(@D) 			# Creates folders if not present
 
 ###############################################################################
 
-all: bin/$(PROJECT).bin
+###########
+# Targets #
+###########
 
-$(OBJ)%.o: src/%.c
-	@echo "\n[OBJ]	Creating OBJs from src"
-	$(MKDIR)
-	$(CC) -o $@ $^ $(INC) $(CFLAGS)
+all: bin/$(PROGRAM_NAME).bin
 
-$(OBJ)%.o: lib/%.c
-	@echo "\n[OBJ]	Creating OBJs from libs"
+$(OBJ_DIR)/%.o: $(SRC_DIR)/%.c
+	@echo "\n[OBJ]	Creating OBJs from src"
 	$(MKDIR)
-	$(CC) -o $@ $^ $(INC) $(CFLAGS)
+	$(CC) -o $@ $^ -I$(INC_DIR) $(LIB_INC_LIST) $(CFLAGS)
 
-obj/$(STARTUP).o: $(SOC_SW_ROOT_DIR)/startup/$(STARTUP).s
-	@echo "\n[OBJ]	Creating OBJs from $(STARTUP).s"
+obj/startup.o: 
```

Comment:
if startup is the same for all bare-metal SoC programs, also this target should be under a common/ folder

Reply:
it is actually better to have it compiled and not shared, since some applications might use different extensions or XLEN to compile the source code.